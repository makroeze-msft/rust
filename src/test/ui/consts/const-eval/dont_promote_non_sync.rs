#![feature(optin_builtin_traits)]

struct Foo;
impl !Sync for Foo {}

struct Bar(i32);
impl !Sync for Bar {}

struct Baz { field: i32 }
impl !Sync for Baz {}

enum Bla { T(i32), S { field: i32 } }
impl !Sync for Bla {}

// Known values of the given types
fn mk_foo() -> &'static Foo { &Foo } //~ ERROR does not live long enough
fn mk_bar() -> &'static Bar { &Bar(0) } //~ ERROR does not live long enough
fn mk_baz() -> &'static Baz { &Baz { field: 0 } } //~ ERROR does not live long enough
fn mk_bla_t() -> &'static Bla { &Bla::T(0) } //~ ERROR does not live long enough
fn mk_bla_s() -> &'static Bla { &Bla::S { field: 0 } } //~ ERROR does not live long enough

// Unknown values of the given types (test a ZST and a non-ZST)
trait FooT { const C: Foo; }
fn mk_foo2<T: FooT>() -> &'static Foo { &T::C } //~ ERROR does not live long enough

trait BarT { const C: Bar; }
fn mk_bar2<T: BarT>() -> &'static Bar { &T::C } //~ ERROR does not live long enough

fn main() {}
