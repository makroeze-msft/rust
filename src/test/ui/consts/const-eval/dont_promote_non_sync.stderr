error[E0597]: borrowed value does not live long enough
  --> $DIR/dont_promote_non_sync.rs:16:32
   |
LL | fn mk_foo() -> &'static Foo { &Foo } //~ ERROR does not live long enough
   |                                ^^^ - temporary value only lives until here
   |                                |
   |                                temporary value does not live long enough
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough
  --> $DIR/dont_promote_non_sync.rs:17:32
   |
LL | fn mk_bar() -> &'static Bar { &Bar(0) } //~ ERROR does not live long enough
   |                                ^^^^^^ - temporary value only lives until here
   |                                |
   |                                temporary value does not live long enough
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough
  --> $DIR/dont_promote_non_sync.rs:18:32
   |
LL | fn mk_baz() -> &'static Baz { &Baz { field: 0 } } //~ ERROR does not live long enough
   |                                ^^^^^^^^^^^^^^^^ - temporary value only lives until here
   |                                |
   |                                temporary value does not live long enough
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough
  --> $DIR/dont_promote_non_sync.rs:19:34
   |
LL | fn mk_bla_t() -> &'static Bla { &Bla::T(0) } //~ ERROR does not live long enough
   |                                  ^^^^^^^^^ - temporary value only lives until here
   |                                  |
   |                                  temporary value does not live long enough
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough
  --> $DIR/dont_promote_non_sync.rs:20:34
   |
LL | fn mk_bla_s() -> &'static Bla { &Bla::S { field: 0 } } //~ ERROR does not live long enough
   |                                  ^^^^^^^^^^^^^^^^^^^ - temporary value only lives until here
   |                                  |
   |                                  temporary value does not live long enough
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough
  --> $DIR/dont_promote_non_sync.rs:24:42
   |
LL | fn mk_foo2<T: FooT>() -> &'static Foo { &T::C } //~ ERROR does not live long enough
   |                                          ^^^^ - temporary value only lives until here
   |                                          |
   |                                          temporary value does not live long enough
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough
  --> $DIR/dont_promote_non_sync.rs:27:42
   |
LL | fn mk_bar2<T: BarT>() -> &'static Bar { &T::C } //~ ERROR does not live long enough
   |                                          ^^^^ - temporary value only lives until here
   |                                          |
   |                                          temporary value does not live long enough
   |
   = note: borrowed value must be valid for the static lifetime...

error: aborting due to 7 previous errors

For more information about this error, try `rustc --explain E0597`.
